File: src/main/java/net/ndrei/teslacorelib/netsync/SimpleNBTHandler.java
package net.ndrei.teslacorelib.netsync;

import net.minecraft.client.Minecraft;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.world.WorldServer;
import net.minecraftforge.common.DimensionManager;
import net.minecraftforge.fml.common.network.simpleimpl.IMessageHandler;
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext;

/**
 * Created by CF on 2016-12-03.
 */
public class SimpleNBTHandler implements IMessageHandler<SimpleNBTMessage, SimpleNBTMessage> {
    @Override
    public SimpleNBTMessage onMessage(SimpleNBTMessage message, MessageContext ctx) {
        if (ctx.side.isClient()) {
            // process client side message
            if ((message != null) && (message.getPos() != null) && (Minecraft.getMinecraft().world != null)) {
                TileEntity entity = Minecraft.getMinecraft().world.getTileEntity(message.getPos());
                if ((entity != null) && (entity instanceof ISimpleNBTMessageHandler)) {
                    return ((ISimpleNBTMessageHandler)entity).handleMessage(message);
                }
            }
        } else {
            // processing server side message
            if (message != null) {
                WorldServer world = DimensionManager.getWorld(message.getDimension());
                if (world != null) {
                    TileEntity entity = world.getTileEntity(message.getPos());
                    if ((entity != null) && (entity instanceof ISimpleNBTMessageHandler)) {
                        return ((ISimpleNBTMessageHandler)entity).handleMessage(message);
                    }
                }
            }
        }
        return null;
    }
}

File: src/main/kotlin/net/ndrei/teslacorelib/netsync/SimpleNBTHandler.kt
package net.ndrei.teslacorelib.netsync

import net.minecraft.client.Minecraft
import net.minecraftforge.common.DimensionManager
import net.minecraftforge.fml.common.network.simpleimpl.IMessageHandler
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext

/**
 * Created by CF on 2017-06-28.
 */
class SimpleNBTHandler : IMessageHandler<SimpleNBTMessage, SimpleNBTMessage> {
    override fun onMessage(message: SimpleNBTMessage?, ctx: MessageContext): SimpleNBTMessage? {
        if (ctx.side.isClient) {
            // process client side message
            if ((message != null) && (message.pos != null) && (Minecraft.getMinecraft().world != null)) {
                val entity = Minecraft.getMinecraft().world.getTileEntity(message.pos)
                if (entity != null && entity is ISimpleNBTMessageHandler) {
                    return (entity as ISimpleNBTMessageHandler).handleServerMessage(message)
                }
            }
        } else {
            // processing server side message
            if ((message != null) && (message.dimension != null) && (message.pos != null)) {
                val world = DimensionManager.getWorld(message.dimension!!)
                if (world != null) {
                    val entity = world.getTileEntity(message.pos!!)
                    if ((entity != null) && (entity is ISimpleNBTMessageHandler)) {
                        return (entity as ISimpleNBTMessageHandler).handleClientMessage(ctx.serverHandler.player, message)
                    }
                }
            }
        }
        return null
    }
}